doc/mpls/MPLS_hld.md

```diff
@@ -0,0 +1,563 @@
+ifname        = STRING           ; Comma-separated list of interfaces.
+```
+
+#### Software Modules
+This section describes modifications to SONiC infrastructure software modules to support MPLS.
+
+##### NetLink
+Netlink is an existing open source library imported to SONiC.
+Modifications to the existing NetLink MPLS implementation were needed to support MPLS attributes.
+###### Functions
+New accessors were added to retrieve the MPLS NH destination and TTL values:
+```
+  /* Accessor to retrieve MPLS destination */
+  extern struct nl_addr *	rtnl_route_nh_get_encap_mpls_dst(struct rtnl_nexthop *);
+  /* Accessor to retrieve MPLS TTL */
+  extern uint8_t		rtnl_route_nh_get_encap_mpls_ttl(struct rtnl_nexthop *);
+```
+
+##### IntfMgr
+IntfMgr is an existing daemon in SWSS container that monitors operations in CONFIG_DB on INTERFACE and PORTCHANNEL_INTERFACE tables.
+
+For MPLS, IntfMgr is modified to additionally process the "mpls" attribute and propagate this attribute to APPL_DB.
+
+##### FPM Syncd
+FPM Syncd is an existing daemon in BGP container that monitors NetLink messages from the SONiC routing stack for Route and Next Hop information.
+
+New support has been added to FPM Syncd for MPLS to process MPLS related Route and Next Hop information in the received NetLink messages and propagate this information to the APPL_DB.
+###### Functions
+The following are new functions for fpmsyncd:
+```
+  /* Handler for rtnl messages with MPLS route */
+  void RouteSync::onLabelRouteMsg(int nlmsg_type, struct nl_object *obj);
+```
+
+##### IntfsOrch
+IntfsOrch is an existing component of the OrchAgent daemon in the SWSS container.  IntfsOrch monitors operations on Interface related tables in APPL_DB and converts those operations into SAI commands to manage the Router Interface object.
+
+For MPLS, IntfsOrch has been extended to detect the new per-RIF MPLS enable/disable attribute in the APPL_DB and propagate this attribute to the ASIC_DB via SAI_ROUTER_INTERFACE_ATTR_ADMIN_MPLS_STATE.  This MPLS behavior parallels the existing IntfsOrch behavior for IPv4/IPv6 with SAI_ROUTER_INTERFACE_ATTR_ADMIN_V4_STATE and SAI_ROUTER_INTERFACE_ATTR_ADMIN_V6_STATE.
+###### Functions
+The following are new functions for IntfsOrch:
+```
+  /* Handler to enable/disable MPLS per Interface */
+  bool IntfsOrch::setRouterIntfMpls(Port& port)
+```
+
+##### RouteOrch
+RouteOrch is an existing component of the OrchAgent daemon in the SWSS container.  RouteOrch monitors operations on Route related tables in APPL_DB and converts those operations in SAI commands to manage Route and In-Segment entries.  Additionally RouteOrch coordinates Next Hop object operations with NeighOrch and converts operations into SAI commands to manage Next Hop Group objects.
+
+For MPLS, RouteOrch was modified to monitor updates to the new APPL_DB LABEL_ROUTE_TABLE.  RouteOrch translates all updates to LABEL_ROUTE_TABLE to equivalent SAI requests for SAI MPLS API.
```

Comment:
Why do we overload RouteOrch? Can we have a separate `LabelRouteOrch`?

Reply:
@prsunny That seems like overkill.  The functionality is identical except for the lookup key, so the grouping is natural.

Reply:
I suggest we separate the code for mpls from routeorch.cpp file.